File: backend/src/services/reportService.js
'use strict';

const reportRepo = require('../repositories/reportRepo');

/**
 * Builds the manager Reports & Analytics tables — Node port of ReportService.
 * Each report has a stable key so the CSV-export endpoint can request exactly one.
 * Returns { key, title, headers, rows }.
 */
const DEFS = [
  { key: 'claims-volume', title: 'Claims Volume by Status', headers: ['Status', 'Claims', 'Estimated (₹)'], fn: reportRepo.claimsByStatus },
  { key: 'claims-type', title: 'Claims by Type', headers: ['Type', 'Claims', 'Estimated (₹)'], fn: reportRepo.claimsByType },
  { key: 'sla-compliance', title: 'SLA Compliance', headers: ['Bucket', 'Claims'], fn: reportRepo.slaCompliance },
  { key: 'settlement-tat', title: 'Settlement TAT', headers: ['Metric', 'Value'], fn: reportRepo.settlementTat },
  { key: 'fraud-detection', title: 'Fraud Detection (high risk)', headers: ['Claim No.', 'Type', 'Risk', 'Fraud Score', 'Status'], fn: reportRepo.fraudWatch },
  { key: 'agent-performance', title: 'Agent Performance', headers: ['Agent', 'Total', 'Settled', 'Pending Approval'], fn: reportRepo.agentPerformance },
];

async function allReports() {
  const out = [];
  for (const d of DEFS) {
    out.push({ key: d.key, title: d.title, headers: d.headers, rows: await d.fn() });
  }
  return out;
}

async function report(key) {
  const d = DEFS.find((x) => x.key === key);
  if (!d) return null;
  return { key: d.key, title: d.title, headers: d.headers, rows: await d.fn() };
}

module.exports = { allReports, report };

File: src/main/java/com/sgx/icms/service/ReportService.java
package com.sgx.icms.service;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import com.sgx.icms.dao.ReportDao;
import com.sgx.icms.db.Db;
import com.sgx.icms.web.support.ReportTable;

/**
 * Builds the manager Reports &amp; Analytics tables. Each report has a stable key so
 * the CSV-export endpoint can request exactly one.
 */
public class ReportService {

    private final ReportDao reportDao = new ReportDao();

    /** All report tables, in display order, keyed by their export key. */
    public Map<String, ReportTable> allReports() {
        Map<String, ReportTable> reports = new LinkedHashMap<>();
        Db.withConnection(conn -> {
            reports.put("claims-volume", new ReportTable("claims-volume", "Claims Volume by Status",
                    Arrays.asList("Status", "Claims", "Estimated (₹)"), rows(reportDao.claimsByStatus(conn))));
            reports.put("claims-type", new ReportTable("claims-type", "Claims by Type",
                    Arrays.asList("Type", "Claims", "Estimated (₹)"), rows(reportDao.claimsByType(conn))));
            reports.put("sla-compliance", new ReportTable("sla-compliance", "SLA Compliance",
                    Arrays.asList("Bucket", "Claims"), rows(reportDao.slaCompliance(conn))));
            reports.put("settlement-tat", new ReportTable("settlement-tat", "Settlement TAT",
                    Arrays.asList("Metric", "Value"), rows(reportDao.settlementTat(conn))));
            reports.put("fraud-detection", new ReportTable("fraud-detection", "Fraud Detection (high risk)",
                    Arrays.asList("Claim No.", "Type", "Risk", "Fraud Score", "Status"), rows(reportDao.fraudWatch(conn))));
            reports.put("agent-performance", new ReportTable("agent-performance", "Agent Performance",
                    Arrays.asList("Agent", "Total", "Settled", "Pending Approval"), rows(reportDao.agentPerformance(conn))));
            return null;
        });
        return reports;
    }

    public ReportTable report(String key) {
        return allReports().get(key);
    }

    private static List<List<String>> rows(List<String[]> raw) {
        List<List<String>> out = new ArrayList<>();
        for (String[] r : raw) {
            out.add(Arrays.asList(r));
        }
        return out;
    }
}
